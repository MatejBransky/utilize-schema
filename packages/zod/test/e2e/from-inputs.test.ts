import type { FileInfo } from '@apidevtools/json-schema-ref-parser';
import type { JSONSchema } from '@utilize/json-schema-core';
import { stripExtension } from '@utilize/json-schema-core';
import merge from 'lodash/merge';
import { expect, it } from 'vitest';

import { readdirSync } from 'node:fs';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { getWithCache } from '../http';
import { compile, type CompileOptions } from '../test-utils';

const dir = fileURLToPath(new URL('inputs', import.meta.url));

interface TestCase {
	input: JSONSchema;
	error?: true;
	exclude?: boolean;
	only?: boolean;
	options?: CompileOptions;
}

export function hasOnly() {
	return readdirSync(dir)
		.filter((_) => /^.*\.js$/.test(_))
		.map((_) => require(join(dir, _)))
		.some((_) => _.only);
}

export async function run() {
	// [filename, absolute dirname, contents][]
	const modules = await Promise.all(
		readdirSync(dir)
			.filter((_) => !_.includes('.ignore.') && _.includes('.input.'))
			.map(async (_) => [_, await import(join(dir, _))] as [string, TestCase])
	);

	// exporting `const only=true` will only run that test
	// exporting `const exclude=true` will not run that test
	const only = modules.find((_) => Boolean(_[1].only));

	if (only) {
		runOne(only[1], only[0]);
	} else {
		modules.filter((_) => !_[1].exclude).forEach((_) => runOne(_[1], _[0]));
	}
}

const httpWithCacheResolver = {
	order: 1,
	canRead: /^https?:/i,
	async read({ url }: FileInfo) {
		return await getWithCache(url);
	},
};

function runOne(exports: TestCase, name: string) {
	// log('blue', 'Running test', name)

	const options = merge<CompileOptions, CompileOptions>(
		{
			normalize: {
				fileName: stripExtension(name),
			},
			generate: { importZod: true },
			dereference: {
				cwd: join(fileURLToPath(new URL(import.meta.url)), 'inputs'),
				$refOptions: {
					resolve: { http: httpWithCacheResolver },
				},
			},
		},
		exports.options ?? {}
	);

	it(name, async () => {
		if (exports.error) {
			try {
				await compile(exports.input, options);
			} catch (e) {
				expect(e instanceof Error).toBe(true);
			}
		} else {
			expect(await compile(exports.input, options)).toMatchSnapshot();
		}
	});
}

await run();
