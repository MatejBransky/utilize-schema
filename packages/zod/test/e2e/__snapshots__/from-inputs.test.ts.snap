// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`All.input.ts 1`] = `
"export const All = z.object({
	allOf: z.intersection(z.boolean(), z.number()).optional(),
	anyOf: z.union([z.boolean(), z.number(), z.string()]).optional(),
	oneOf: z.union([z.boolean(), z.number(), z.string()]).optional(),
	array: z.array(z.string()).min(2).max(3).optional(),
	tuple: z.tuple([z.boolean(), z.number(), z.string()]).optional(),
	const: z.literal('xbox').optional(),
	enum: z.enum(['ps4', 'ps5']).optional(),
	ifThenElse: z.object({}).optional(),
	null: z.null().optional(),
	multiple: z.union([z.array(z.unknown()), z.boolean()]).optional(),
	objAdditionalTrue: z
		.object({
			x: z.string().optional(),
		})
		.catchall(z.unknown())
		.optional(),
	objAdditionalFalse: z
		.strictObject({
			x: z.string().optional(),
		})
		.optional(),
	objAdditionalNumber: z
		.object({
			x: z.string().optional(),
		})
		.catchall(z.number())
		.optional(),
	objAdditionalOnly: z.object({}).catchall(z.number()).optional(),
	patternProps: z
		.strictObject({
			z: z.string().optional(),
			'^x': z.string().optional(),
			'^y': z.number().optional(),
		})
		.optional(),
});
export type All = z.infer<typeof All>;
"
`;

exports[`Step.input.ts 1`] = `""`;
